src/unix/futex.c

```diff
@@ -294,3 +294,106 @@ init_futices(process p)
         halt("failed to allocate futex table\n");
 
 }
+
+/* robust mutex handling */
+
+#define FUTEX_OWNER_DIED	0x40000000
+#define FUTEX_KEY_ADDR(x, o)    ((u32 *)((u8 *)(x) + (o)))
+
+typedef struct robust_list {
+    struct robust_list *next;
+} *robust_list;
+
+typedef struct robust_list_head {
+    robust_list list;
+    long futex_offset;
+    void *list_op_pending;
+} *robust_list_head;
+
+static boolean valid_user_address(process p, void *a, int len)
+{
+    u64 pg = u64_from_pointer(a) & ~PAGEMASK;
+    u64 end = (pg + len) & ~PAGEMASK;
+
+    for (; pg <= end; pg += PAGESIZE) {
+        if (vmap_from_vaddr(p, pg) == INVALID_ADDRESS)
+                return false;
+    }
+    return true;
+}
+
+void wake_robust_list(process p, void *head)
+{
+    struct futex * f;
+    struct robust_list_head *h = head;
+    struct robust_list *l;
+    u32 *uaddr;
+
+    /* must be very careful accessing the head as well as the list */
+    if (!valid_user_address(p, h, sizeof *h))
+        return;
+
+    /* XXX could keep a list of futexes and wake them at the end
+     * to let threads acquire multiple locks without blocking */
+    if (h->list_op_pending) {
+        uaddr = FUTEX_KEY_ADDR(h->list_op_pending, h->futex_offset);
+        if (valid_user_address(p, uaddr, sizeof *uaddr)) {
+            *uaddr |= FUTEX_OWNER_DIED;
+            if ((f = table_find(p->futices, uaddr)))
+                futex_wake_many(f, 1);
+        }
+    }
+
+    for (l = h->list; (void *)l != (void *)h; l = l->next) {
+        uaddr = FUTEX_KEY_ADDR(l, h->futex_offset);
+        if (!valid_user_address(p, l, sizeof *l))
+            break;
+        if (!valid_user_address(p, uaddr, sizeof *uaddr))
+            break;
+        *uaddr |= FUTEX_OWNER_DIED;
+        if ((f = table_find(p->futices, uaddr)))
+            futex_wake_many(f, 1);
+    }
+}
+
+sysreturn get_robust_list(int pid, void *head, u64 *len)
+{
+    struct robust_list_head **hp = head;
+    if (!validate_user_memory(hp, sizeof *hp, true))
+        return set_syscall_error(current, EFAULT);
```

Comment:
I know that this is being done in other syscalls too, but it's a bit wasteful to call `set_syscall_error()` (or `set_syscall_return()`) in syscalls that return directly without blocking, because the return value is set in syscall_debug(), so I think in new code we should avoid this double setting and just return the value (with minus sign if it's an error).

Reply:
Good point, thanks!